L8-c++-internalization.md

```diff
@@ -0,0 +1,172 @@
+Separate internal-only and public sections of the C++ wrapping
+----
+* Author(s): vjpai
+* Approver: ctiller
+* Status: Approved
+* Implemented in: C++
+* Last updated: July 6, 2017
+* Discussion at: https://groups.google.com/forum/#!topic/grpc-io/Z8S9P4NzGHs
+
+## Abstract
+
+Maintain separation of C++ classes and functions that are intended
+to be directly used by application developers from those that are
+supposed to be used only through codegen or for that exist for gRPC's
+internal implementation (e.g., interactions with the core API).
+
+## Background
+
+Essentially all of gRPC C++ lives in `namespace grpc`, giving the
+appearance that all portions are equally intended as part of the
+publicly-exposed API. This gRFC aims to clarify this by categorizing
+the gRPC C++ wrapping, moving some components to `namespace
+grpc::internal`, and moving some class constructors to `private` (with
+friends or appropriately-named `static` factories for internalized generation).
```

Comment:
I think that the template issue has been exactly our problem in the past: we have said that `include/grpc++/impl` is not public but then many headers in `include/grpc++` reach into that to provide the needed templates. Our current feeling (as indicated in our [BUILD](https://github.com/grpc/grpc/blob/master/BUILD) files, for example) is that anything in `include/grpc++` is ultimately public but we can make it clear what is and isn't really meant to be used by applications by spelling out the internalization directly.